src/macros.jl

```diff
@@ -474,3 +474,30 @@ for file in readdir(joinpath(@__DIR__, "macros"))
         include(joinpath(@__DIR__, "macros", file))
     end
 end
+
+# These methods must come after the macros are included, because they use both
+# `@variable` and `@constraint`.
+
+function _build_subexpression(
+    ::Function,
+    model::AbstractModel,
+    expr::AbstractJuMPScalar,
+    name::String,
+)
+    y = @variable(model)
+    set_name(y, name)
+    @constraint(model, y == expr)
+    return y
```

Comment:
```suggestion
    return y, expr
```
This would enable
```julia
julia> model = Model()
A JuMP Model
├ solver: none
├ objective_sense: FEASIBILITY_SENSE
├ num_variables: 0
├ num_constraints: 0
└ Names registered in the model: none

julia> @variable(model, x)
x

julia> @expression(model, ex, sin(x), subexpression = true)
(ex, sin(x))

julia> ex
(ex, sin(x))

julia> model[:ex]
(ex, sin(x))

julia> model = Model()
A JuMP Model
├ solver: none
├ objective_sense: FEASIBILITY_SENSE
├ num_variables: 0
├ num_constraints: 0
└ Names registered in the model: none

julia> @variable(model, x[1:2])
2-element Vector{VariableRef}:
 x[1]
 x[2]

julia> @expression(model, ex[i in 1:2], sin(x[i]), subexpression = true)
2-element Vector{Tuple{VariableRef, NonlinearExpr}}:
 (ex[1], sin(x[1]))
 (ex[2], sin(x[2]))

julia> first.(ex)
2-element Vector{VariableRef}:
 ex[1]
 ex[2]

julia> last.(ex)
2-element Vector{NonlinearExpr}:
 sin(x[1])
 sin(x[2])
```

which seems quite nice.